CRM/Case/BAO/Case.php

```diff
@@ -1176,9 +1176,11 @@ public static function getRelatedContacts($caseID, $includeDetails = TRUE) {
      AND cr.is_active
      AND cc.id NOT IN (%2)
 HERESQL;
+
+    $clientIdType = !empty($caseInfo['client_id']) ? 'CommaSeparatedIntegers' : 'String';
     $params = [
       1 => [$caseID, 'Integer'],
-      2 => [implode(',', $caseInfo['client_id']), 'String'],
+      2 => [implode(',', $caseInfo['client_id']), $clientIdType],
```

Comment:
It looks to me like you're also trying to handle the scenario where `$caseInfo['client_id']` is an empty string. But in that case it would get injected into a sql query like `AND cc.id NOT IN ("")` which is bad if not invalid SQL. Is that a real possibility? In that case we'd need to handle it better, probably by altering the query to omit that clause.

Reply:
@colemanw , It is a possibility, when I changed the param type to `CommaSeparatedIntegers` directly, I got an error emanating from the civirules module. I described the error under the Proposed solution in this section here: https://lab.civicrm.org/dev/core/issues/1659, so the issue had always been there and the query will run apparently when `$caseInfo['client_id']` is an empty string but it's bad SQL.
 
I can update the PR to check for this and exclude the condition from the SQL and also add a test. Can you point me to a test similar to what I need to add for this?

Reply:
Fortunately, that clause is literally the last line in the query. So you could delete that line from the query, and also delete key `2` from the `$params` array, then do a conditional afterwards:

    if (!empty($caseInfo['client_id'])) {
      $query .= <<<HERESQL
    AND cc.id NOT IN (%2)
    HERESQL;
      $params[2] = [implode(',', $caseInfo['client_id']), 'CommaSeparatedIntegers'];
    }

Reply:
@colemanw 
The clause is actually in two places in the Query. Here https://github.com/civicrm/civicrm-core/pull/16837 and https://github.com/civicrm/civicrm-core/pull/16837. 

Was thinking of fixing like this:
```php
$caseClientCondition = !empty($caseInfo['client_id']) ? "AND cc.id NOT IN (%2)" : '';
    $query = <<<HERESQL
    SELECT cc.display_name as name, cc.sort_name as sort_name, cc.id, cr.relationship_type_id, crt.label_b_a as role, crt.name_b_a as role_name, ce.email, cp.phone
    FROM civicrm_relationship cr
    JOIN civicrm_relationship_type crt
     ON crt.id = cr.relationship_type_id
    JOIN civicrm_contact cc
     ON cc.id = cr.contact_id_a
     AND cc.is_deleted <> 1
    LEFT JOIN civicrm_email ce
     ON ce.contact_id = cc.id
     AND ce.is_primary= 1
    LEFT JOIN civicrm_phone cp
     ON cp.contact_id = cc.id
     AND cp.is_primary= 1
    WHERE cr.case_id =  %1
     AND cr.is_active
     {$caseClientCondition}
    UNION
    SELECT cc.display_name as name, cc.sort_name as sort_name, cc.id, cr.relationship_type_id, crt.label_a_b as role, crt.name_a_b as role_name, ce.email, cp.phone
    FROM civicrm_relationship cr
    JOIN civicrm_relationship_type crt
     ON crt.id = cr.relationship_type_id
    JOIN civicrm_contact cc
     ON cc.id = cr.contact_id_b
     AND cc.is_deleted <> 1
    LEFT JOIN civicrm_email ce
     ON ce.contact_id = cc.id
     AND ce.is_primary= 1
    LEFT JOIN civicrm_phone cp
     ON cp.contact_id = cc.id
     AND cp.is_primary= 1
    WHERE cr.case_id =  %1
     AND cr.is_active
     {$caseClientCondition}
HERESQL;

   if ($caseClientCondition) {
      $params[2] = array(implode(',', $caseInfo['client_id']), 'CommaSeparatedIntegers');
    }
```

Reply:
That makes sense